nexus_motion_planner/src/motion_plan_cache.hpp

```diff
@@ -0,0 +1,138 @@
+ *     - Final robot joint states
+ *
+ * Motion plans may be looked up with some tolerance.
+ */
+class MotionPlanCache
+{
+public:
+  // We explicitly need a Node::SharedPtr because warehouse_ros ONLY supports
+  // it...
+  explicit MotionPlanCache(const rclcpp::Node::SharedPtr& node);
+
+  void init(
+    const std::string& db_path = ":memory:",
+    uint32_t db_port = 0,
+    double exact_match_precision = 0.00001);
+
+  // TOP LEVEL OPS =============================================================
+  std::vector<
+    warehouse_ros::MessageWithMetadata<
+      moveit_msgs::msg::RobotTrajectory
+    >::ConstPtr
+  >
+  fetch_all_matching_plans(
+    const moveit::planning_interface::MoveGroupInterface& move_group,
+    const std::string& move_group_namespace,
+    const moveit_msgs::msg::MotionPlanRequest& plan_request,
+    double start_tolerance, double goal_tolerance, bool metadata_only = false);
+
+  warehouse_ros::MessageWithMetadata<
+    moveit_msgs::msg::RobotTrajectory
+  >::ConstPtr
+  fetch_best_matching_plan(
+    const moveit::planning_interface::MoveGroupInterface& move_group,
+    const std::string& move_group_namespace,
+    const moveit_msgs::msg::MotionPlanRequest& plan_request,
+    double start_tolerance, double goal_tolerance, bool metadata_only = false);
+
+  bool put_plan(
+    const moveit::planning_interface::MoveGroupInterface& move_group,
+    const std::string& move_group_namespace,
+    const moveit_msgs::msg::MotionPlanRequest& plan_request,
+    const moveit_msgs::msg::RobotTrajectory& plan,
+    double execution_time_s,
+    double planning_time_s,
+    bool overwrite = true);
+
+  // QUERY CONSTRUCTION ========================================================
+  bool extract_and_append_start_to_query(
+    warehouse_ros::Query& query,
+    const moveit::planning_interface::MoveGroupInterface& move_group,
+    const moveit_msgs::msg::MotionPlanRequest& plan_request,
+    double match_tolerance);
+
+  bool extract_and_append_goal_to_query(
+    warehouse_ros::Query& query,
+    const moveit::planning_interface::MoveGroupInterface& move_group,
+    const moveit_msgs::msg::MotionPlanRequest& plan_request,
+    double match_tolerance);
+
+  // METADATA CONSTRUCTION =====================================================
+  bool extract_and_append_start_to_metadata(
+    warehouse_ros::Metadata& metadata,
+    const moveit::planning_interface::MoveGroupInterface& move_group,
+    const moveit_msgs::msg::MotionPlanRequest& plan_request);
+
+  bool extract_and_append_goal_to_metadata(
+    warehouse_ros::Metadata& metadata,
+    const moveit::planning_interface::MoveGroupInterface& move_group,
+    const moveit_msgs::msg::MotionPlanRequest& plan_request);
+
+private:
+  rclcpp::Node::SharedPtr node_;
+  warehouse_ros::DatabaseConnection::Ptr db_;
+
+  double exact_match_precision_ = 0;
+
+  std::unique_ptr<tf2_ros::Buffer> tf_buffer_;
+  std::shared_ptr<tf2_ros::TransformListener> tf_listener_;
```

Comment:
The tf is just to pull the state of the robot at time of planning (to then explicitly fill into the cache when adding the cache entry.) (It is also to pull the state of the goal pose wrt. the planning frame (`base_link`) and explicitly save it.

This is because there are certain motion plan requests that call for the current state of the robot (without actually telling the planner what the current state of the robot is, so it has to look it up.) (Same with the goal.)

See the `is_diff=true`, which requires getting the current state` variant of motion plan request.

In all of these cases (where plan request entries are "dynamic", the caching class I implement explicitly looks up the values to statically instantiate them as joint states or poses relative to a common, unchanging frame (`base_link`, or whatever was defined as the workspace frame.)

At no point in time are cache entries edited after they are inserted (unless they are deleted because a better entry was found.)

Reply:
e.g.

A goal constraint:
```
  position_constraints:
  -
    header:
      stamp:
        sec: 0
        nanosec: 0
      frame_id: "pickup_label_goal_pose"
    link_name: "gripper_link"
    target_point_offset:
      x: 0.00000
      y: 0.00000
      z: 0.00000
```

Will be converted at plan time into:
```
```
  position_constraints:
  -
    header:
      stamp:
        sec: 0
        nanosec: 0
      frame_id: "base_link"
    link_name: "gripper_link"
    target_point_offset:
      x: SOME_NUM
      y: SOME_NUM
      z: SOME_NUM
```

Where the pose is the transform from `base_link` to `pickup_label_goal_pose` at time of motion plan request.